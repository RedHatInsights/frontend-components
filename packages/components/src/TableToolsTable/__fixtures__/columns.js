import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
  },
  {
    title: 'Name',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
];
