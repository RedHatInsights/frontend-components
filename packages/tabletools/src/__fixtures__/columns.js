import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
    key: 'id',
    managable: false,
  },
  {
    title: 'Name',
    key: 'id',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    key: 'another_name_column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    key: 'desc',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
];
