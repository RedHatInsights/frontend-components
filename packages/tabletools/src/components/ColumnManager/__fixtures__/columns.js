export default [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: 'Another col',
    key: 'name',
    isRequired: true,
  },
  {
    title: 'Name',
    key: 'name1',
  },
  {
    title: 'Another Name column without key',
  },
];
