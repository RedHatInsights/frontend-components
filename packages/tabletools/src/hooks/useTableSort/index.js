export { default } from './useTableSort';
export { default as useTableSortWithItems } from './useTableSortWithItems';

export * from './useTableSort';
