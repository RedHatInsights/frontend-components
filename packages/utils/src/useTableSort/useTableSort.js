import { useState } from 'react';
import { columnOffset, addSortableTransform } from './helpers';

const useTableSort = (columns, options = {}) => {
  const [sortBy, setSortBy] = useState(
    options.sortBy || {
      index: columnOffset(options),
      direction: 'asc',
    }
  );
  const onSort = (_, index, direction) =>
    setSortBy({
      index,
      direction,
    });

  return {
    sortBy,
    tableProps: {
      onSort,
      sortBy,
      cells: addSortableTransform(columns),
    },
  };
};

export default useTableSort;
