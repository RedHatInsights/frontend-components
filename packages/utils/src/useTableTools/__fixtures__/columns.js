import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        key: 'name',
        sortByProperty: 'name',
        renderFunc: (item) => (
            item.name
        )
    }, {
        title: 'Description',
        key: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (severityLevels.indexOf(item.severity)),
        renderFunc: (item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (item) => (
            item.severity
        )
    }
];
