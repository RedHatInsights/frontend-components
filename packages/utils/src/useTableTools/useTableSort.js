import { useState } from 'react';
import { sortable } from '@patternfly/react-table';

const getSortable = (property, item) => {
    if (typeof(property) === 'function') {
        return property(item);
    } else {
        return item[property];
    }
};

const compareSortableAsStrings = (property, a, b) => (
    String(getSortable(property, a))
    .localeCompare(String(getSortable(property, b)))
);

export const orderArrayByProp = (property, objects, direction) => (
    objects.sort((a, b) => {
        if (direction === 'asc') {
            return compareSortableAsStrings(property, a, b);
        } else {
            return -compareSortableAsStrings(property, a, b);
        }
    })
);
export const orderByArray = (objectArray, orderProp, orderArray, direction) => (
    (direction !== 'asc' ? orderArray.reverse() : orderArray).flatMap((orderKey) => (
        objectArray.filter((item) => (item[orderProp] === orderKey))
    ))
);

const isSortable = (column) => (
    column.sortByProp || column.sortByFunction
);

const addSortableTransform = (columns) => (
    columns.map((column) => ({
        ...column,
        ...isSortable(column) ? { transforms: Array.from(new Set([
            ...(column?.transforms || []),
            sortable
        ])) } : {}
    }))
);

const columnOffset = (options = {}) => {
    let offset = 0;

    if (options.selectable) {
        offset += 1;
    }

    if (typeof options.detailsComponent  !== 'undefined') {
        offset += 1;
    }

    return offset;
};

const sorter = (columns, options, sortBy) => (
    (items) => {
        const offset = columnOffset(options);
        const column = columns[sortBy.index - offset];
        const sortedItems = column?.sortByArray ? orderByArray(
            items, column?.sortByProperty, column?.sortByArray, sortBy.direction
        ) : orderArrayByProp(
            (column?.sortByProperty || column?.sortByFunction), items, sortBy.direction
        );

        return sortedItems;
    }
);

const useTableSort = (columns, options = {}) => {
    const [ sortBy, setSortBy ] = useState({
        index: 0,
        direction: 'desc'
    });
    const onSort = (_, index, direction) => {
        setSortBy({
            index,
            direction
        });
    };

    return {
        sorter: sorter(columns, options, sortBy),
        tableProps: {
            onSort,
            sortBy
        },
        columns: addSortableTransform(columns)
    };
};

export default useTableSort;
