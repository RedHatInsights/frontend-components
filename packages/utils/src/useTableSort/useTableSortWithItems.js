import { orderArrayByProp, orderByArray, columnOffset } from './helpers';
import useTableSort from './useTableSort';

const useTableSortWithItems = (items, columns, options) => {
  const { tableProps, sortBy } = useTableSort(columns, options);
  const currentSortableColumn = columns[sortBy.index - columnOffset(options)];

  const sorter = (items) =>
    currentSortableColumn?.sortByArray
      ? orderByArray(items, currentSortableColumn?.sortByProp, currentSortableColumn?.sortByArray, sortBy.direction)
      : orderArrayByProp(currentSortableColumn?.sortByProp || currentSortableColumn?.sortByFunction, items, sortBy.direction);

  return {
    sortBy,
    tableProps: {
      ...tableProps,
      sortBy: items.length > 0 ? sortBy : undefined,
    },
    sorter,
  };
};

export default useTableSortWithItems;
