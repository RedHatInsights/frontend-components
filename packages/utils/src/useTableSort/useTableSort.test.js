import { renderHook } from '@testing-library/react-hooks';
import useTableSort from './useTableSort';
import columns from '../__fixtures__/columns';

describe('useTableSort', () => {
  it('returns a table sort configuration', () => {
    const { result } = renderHook(() => useTableSort(columns));
    expect(result).toMatchSnapshot();
  });

  it('returns a table sort configuration with an inital state', () => {
    const sortBy = {
      index: 3,
      direction: 'asc',
    };
    const { result } = renderHook(() =>
      useTableSort(columns, {
        sortBy,
      })
    );
    expect(result.current.tableProps.sortBy).toEqual(sortBy);
  });
});
