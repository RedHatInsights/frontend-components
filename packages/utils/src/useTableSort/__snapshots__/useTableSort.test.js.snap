// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableSort returns a table sort configuration 1`] = `
Object {
  "all": Array [
    Object {
      "sortBy": Object {
        "direction": "asc",
        "index": 0,
      },
      "tableProps": Object {
        "cells": Array [
          Object {
            "exportKey": "id",
            "key": "id",
            "managable": false,
            "sortByProperty": "id",
            "title": "ID",
          },
          Object {
            "exportKey": "name",
            "key": "id",
            "renderFunc": [Function],
            "sortByProperty": "name",
            "title": "Name",
          },
          Object {
            "exportKey": "name",
            "key": "another_name_column",
            "renderExport": [Function],
            "sortByProperty": "name",
            "title": "Another Name column",
          },
          Object {
            "key": "desc",
            "renderExport": [Function],
            "sortByProperty": "description",
            "title": "Description",
          },
          Object {
            "exportKey": "severity",
            "renderFunc": [Function],
            "sortByFunction": [Function],
            "title": "Severity sorted by function",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "exportKey": "severity",
            "renderFunc": [Function],
            "sortByArray": Array [
              "high",
              "medium",
              "low",
              "unknown",
            ],
            "sortByProperty": "severity",
            "title": "Severity by Array",
          },
        ],
        "onSort": [Function],
        "sortBy": Object {
          "direction": "asc",
          "index": 0,
        },
      },
    },
  ],
  "current": Object {
    "sortBy": Object {
      "direction": "asc",
      "index": 0,
    },
    "tableProps": Object {
      "cells": Array [
        Object {
          "exportKey": "id",
          "key": "id",
          "managable": false,
          "sortByProperty": "id",
          "title": "ID",
        },
        Object {
          "exportKey": "name",
          "key": "id",
          "renderFunc": [Function],
          "sortByProperty": "name",
          "title": "Name",
        },
        Object {
          "exportKey": "name",
          "key": "another_name_column",
          "renderExport": [Function],
          "sortByProperty": "name",
          "title": "Another Name column",
        },
        Object {
          "key": "desc",
          "renderExport": [Function],
          "sortByProperty": "description",
          "title": "Description",
        },
        Object {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByFunction": [Function],
          "title": "Severity sorted by function",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByArray": Array [
            "high",
            "medium",
            "low",
            "unknown",
          ],
          "sortByProperty": "severity",
          "title": "Severity by Array",
        },
      ],
      "onSort": [Function],
      "sortBy": Object {
        "direction": "asc",
        "index": 0,
      },
    },
  },
  "error": undefined,
}
`;
